[mlir][Transform] `apply_conversion_patterns`: Update handles

matthias-springer opened the pull request:
Until now, `transform.apply_conversion_patterns` consumed the target handle and potentially invalidated handles. This commit adds tracking functionality similar to `transform.apply_patterns`, such that handles are no longer invalidated, but updated based on op replacements performed by the dialect conversion.

This new functionality is hidden behind a `preserve_handles` attribute for now.

ftynse commented on mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:
```diff
@@ -190,19 +190,29 @@ def ApplyConversionPatternsOp : TransformDialectOp<"apply_conversion_patterns",
+    a tracking listener that is attached to the dialect conversion, similar to
+    `transform.apply_patterns`. Only replacements via `RewriterBase::replaceOp`
+    or `replaceOpWithNewOp` are considered "payload op replacements". In
+    contrast to `transform.apply_patterns`, we allow replacement ops even if the
```
Nit: could you provide the rationale as to why op name change is allowed here? Presumably because conversion is expected to intentionally change names to another dialect.

ftynse commented on mlir/lib/Dialect/Transform/IR/TransformOps.cpp:
```diff
@@ -632,7 +663,11 @@ LogicalResult transform::ApplyConversionPatternsOp::verify() {
 void transform::ApplyConversionPatternsOp::getEffects(
     SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
-  transform::consumesHandle(getTarget(), effects);
+  if (!getPreserveHandles()) {
+    transform::consumesHandle(getTarget(), effects);
+  } else {
+    transform::onlyReadsHandle(getTarget(), effects);
+  }
```
Nit: I don't recall if dialect conversion could rewrite the top-level op or not. If it can, it may need to still consume the handle...

matthias-springer replied:
Does it actually matter? If the top-level op is rewritten, it will be done through the rewriter, so the transform dialect state will be updated accordingly. (Same as for non-top-level ops.)

matthias-springer committed:
[mlir][Transform] `apply_conversion_patterns`: Update handles

Until now, `transform.apply_conversion_patterns` consumed the target handle and potentially invalidated handles. This commit adds tracking functionality similar to `transform.apply_patterns`, such that handles are no longer invalidated, but updated based on op replacements performed by the dialect conversion.

This new functionality is hidden behind a `preserve_handles` attribute for now.
